refactor STS code

kailun-qin opened the pull request:
Currently, we have two STS clients/plugins that are solving the same problem and
communicating with the same thing (google STS) but all the code is duplicated.
Besides, we have abstractions e.g. `TokenProvider`, `TokenExchanger` and
`TokenManager` which may bring a few confusions sometimes.

This patch tries to clean up and refactor STS code a bit to address the above
concerns.

Fixes https://github.com/istio/istio/issues/29753

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure


Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

istio-testing commented:
@kailun-qin: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
release-notes_istio | 8bdbc29ebbbe4bc5f7781a8e798e448c816ddd20 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/31199/release-notes_istio/1367051634185605120) | `/test release-notes_istio`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

kailun-qin committed:
refactor STS code

Currently, we have two STS clients/plugins that are solving the same problem and
communicating with the same thing (google STS) but all the code is duplicated.
Besides, we have abstractions e.g. `TokenProvider`, `TokenExchanger` and
`TokenManager` which may bring a few confusions sometimes.

This patch tries to clean up and refactor STS code a bit to address the above
concerns.

Fixes https://github.com/istio/istio/issues/29753

Signed-off-by: Kailun Qin <kailun.qin@intel.com>

lei-tang commented on pilot/cmd/pilot-agent/security/security.go:
```diff
@@ -60,10 +58,6 @@ func SetupSecurityOptions(proxyConfig meshconfig.ProxyConfig, secOpt security.Op
 	}
-	// TODO extract this logic out to a plugin
-	if o.CAProviderName == "GoogleCA" || strings.Contains(o.CAEndpoint, "googleapis.com") {
-		o.TokenExchanger = stsclient.NewSecureTokenServiceExchanger(o.CredFetcher, o.TrustDomain)
```
Are the functions in TokenExchanger merged into TokenManager?

lei-tang commented on security/pkg/stsservice/tokenmanager/google/tokenexchangeplugin.go:
```diff
@@ -306,6 +310,38 @@ func (p *Plugin) fetchFederatedToken(parameters security.StsRequestParameters) (
 }
 
+// ExchangeToken exchanges oauth access token from K8s SA jwt.
+func (p *Plugin) ExchangeToken(k8sSAjwt string) (string, error) {
```
Is this function replacing "func (p *SecureTokenServiceExchanger) ExchangeToken(k8sSAjwt string)"? In original ExchangeToken(), there is a retry mechanism. Why is the retry mechanism removed?

func (p *SecureTokenServiceExchanger) ExchangeToken(k8sSAjwt string) (string, error) {
         ....
	body, err := p.requestWithRetry(jsonStr)
         ...
}

lei-tang commented on security/pkg/nodeagent/plugin/providers/google/stsclient/stsclient_test.go:
```diff
@@ -1,92 +0,0 @@
-	"istio.io/istio/tests/util/leak"
-)
-
-func TestGetFederatedToken(t *testing.T) {
```
Where is the equivalent of this test in the refactored code?

lei-tang commented:
The STS code is used in multiple places (e.g., Google CA, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow) and there are subtle differences of its usages in different flows. For example, the Google CA flow directly uses the access token of the federated token whereas other flows call constructGenerateAccessTokenRequest(fResp *federatedTokenResponse) to further exchange the federated token to another access token. 
To reduce the risks of breaking these flows that use STS, the refactoring should have tests to show that these flows are working normally.

lei-tang commented:
Can you describe what changes this refactoring makes?

lei-tang commented:
@kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?

liminw commented:
/hold I want to make sure @JimmyCYJ has carefully reviewed and approved the PR before it is merged.

kailun-qin commented:
> @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?

@lei-tang 
Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
Sure, let me go through the questions and discuss in the next Security WG meeting.

lei-tang commented:
> > @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?
> 
> @lei-tang
> Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
> Sure, let me go through the questions and discuss in the next Security WG meeting.

Thank you! Can you share a design document for review?

kailun-qin commented:
> > > @kailun-qin As STS is crucial to multiple flows (e.g., Google CA flow, telemetry token exchange flow for stackdriver, XDS token exchange flow, and VM token exchange flow), can you present the PR and its design on Istio Security Group Meeting (https://docs.google.com/document/d/12Xz8fCuql2tPL-cpdLXGGE44kYT4o9NB6mljSiS_01U/edit#heading=h.o8pz6aqnzzgk)?
> > 
> > 
> > @lei-tang
> > Excuse me for the late reply (just busy with some internal stuff...) and thanks a lot for all the comments!
> > Sure, let me go through the questions and discuss in the next Security WG meeting.
> 
> Thank you! Can you share a design document for review?

Sure, let me prepare for it. I’ll keep you updated once it’s ready.

istio-testing commented:
@kailun-qin: PR needs rebase.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>
